File: binaryTree.js
let data = []
let result = []

function collect (node, data, depth = 0) {
  if (!node) return false

  if (depth === data.length) data.push([])

  data[depth].push(node.val)

  collect(node.left, data, depth + 1)
  collect(node.right, data, depth + 1)
}

function avgByDepth (node) {
  collect(node, data)

  data.forEach(level => {
    let sum = 0

    level.forEach(number => {
      sum += number
    })

    result.push(sum / level.length)
  })

  return result
}

let n8 = { val: 2, left: false, right: false }
let n7 = { val: 6, left: n8, right: false }
let n6 = { val: 6, left: false, right: false }
let n5 = { val: 2, left: false, right: n7 }
let n4 = { val: 10, left: false, right: false }
let n3 = { val: 9, left: n6, right: false }
let n2 = { val: 7, left: n4, right: n5 }
let n1 = { val: 4, left: n2, right: n3 }

let finalResult = avgByDepth(n1)

console.log(finalResult)

File: binaryTree.py
class Node():
  def __init__(self, val, left, right):
    self.val = val
    self.left = left
    self.right = right

def collect(node, data, depth = 0):
  if not node:
    return None

  if depth not in data:
    data[depth] = []

  data[depth].append(node.val)

  collect(node.left, data, depth + 1)
  collect(node.right, data, depth + 1)

  return None

def avgByDepth(node):
  data = {}
  result = []

  collect(node, data)

  i = 0
  while i in data:
    nums = data[i]
    avg = sum(nums) / len(nums)
    result.append(avg)
    i += 1

  return result

n8 = Node(2, False, False)
n7 = Node(6, n8, False)
n6 = Node(6, False, False)
n5 = Node(2, False, n7)
n4 = Node(10, False, False)
n3 = Node(9, n6, False)
n2 = Node(7, n4, n5)
n1 = Node(4, n2, n3)

finalResult = avgByDepth(n1)

print(finalResult)
